src/gather.jl

```diff
@@ -0,0 +1,82 @@
+export gather, gather!
+
+"""
+    gather!(dst, src, idx)
+
+Reverse operation of scatter, which gather data in `src` to destination according to `idx`.
+For each index `k` in `idx`, assign values to `dst` according to
+
+    dst[:,...,k...] = src[:, ..., idx[k]...]
+
+# Arguments
+- `dst`: the destination where data would be assigned to.
+- `src`: the source data to be assigned.
+- `idx`: the mapping for assignment from source to destination.
+"""
+function gather!(dst::AbstractArray{Tdst,Ndst}, 
+                 src::AbstractArray{Tsrc,Nsrc}, 
+                 idx::AbstractArray{NTuple{M,Int}, Nidx}) where {Tdst,Tsrc,Ndst,Nsrc,Nidx,M}
+  # TODO: use M = _length(eltype(idx)) to merge the integer method into this?
+  @boundscheck _gather_checkbounds(src, idx)
```

Comment:
It would error if it's out of bounds anyway right?

Reply:
no because the loop is marked as inbounds

Reply:
Yes and it shouldnt